staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -744,8 +745,25 @@ func (r *Request) DoRaw() ([]byte, error) {
 func (r *Request) transformResponse(resp *http.Response, req *http.Request) Result {
 	var body []byte
 	if resp.Body != nil {
-		if data, err := ioutil.ReadAll(resp.Body); err == nil {
+		data, err := ioutil.ReadAll(resp.Body)
+		switch err.(type) {
+		case nil:
 			body = data
+		case http2.StreamError:
+			// This is trying to catch the scenario that the server may close the connection when sending the
+			// response body. This can be caused by server timeout due to a slow network connection.
+			// TODO: Add test for this. Steps may be:
+			// 1. client-go (or kubectl) sends a GET request.
+			// 2. Apiserver sends back the headers and then part of the body
+			// 3. Apiserver closes connection.
+			// 4. client-go should catch this and return an error.
+			glog.V(2).Infof("Stream error %#v when reading response body, may be caused by closed connection.", err)
+			streamErr := fmt.Errorf("Stream error %#v when reading response body, may be caused by closed connection. Please retry.", err)
+			return Result{
+				err: streamErr,
+			}
+		default:
+			glog.Errorf("Unexpected error when reading response body: %#v", err)
```

Comment:
Oh, I think we should return `Result { err: fmt.Errorf("Unexpected error...") }` here, too.  Otherwise we still try to decode the truncated "json" and print an error about that.

```
INFO: Running command line: bazel-bin/cmd/kubectl/kubectl '--request-timeout=3000ms' get cm -n kube-system
E0828 16:12:41.403960    7263 round_trippers.go:169] CancelRequest not implemented
E0828 16:12:41.404050    7263 request.go:766] Unexpected error when reading response body: &http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true}
unexpected end of JSON input
```

Reply:
I agree. I will update it soon.
> http.httpError{err:"net/http: request canceled (Client.Timeout exceeded while reading body)", timeout:true}

I'm curious how do you get this error and it ends up have the same error `unexpected end of JSON input`.

Reply:
I chose --request-timeout=3000ms because I was getting headers on my cluster at about 2300ms, and then it would take a long time for the rest of the request to return.  Setting this flag to this value, we get the HTTP 200, but then *we* cause the connection to close, rather than the other side.

But in this case, the same thing happens as before: we log the error and return an uninitialized body.